File: public/waveguide-stress/PackingDensity.java
/**
 * PECVD Ge-doped SiO2 packing density — reference implementation.
 *
 * Three estimators (matching the web page):
 *   A. Lorentz-Lorentz    : from measured refractive index n
 *   B. Density ratio       : p = rhoFilm / rhoBulk
 *   C. Deposition-condition: empirical Arrhenius/oxidation/ion model
 */
public class PackingDensity {

    /** Fused silica Sellmeier index (Malitson 1965). */
    static double nSiO2(double wlUm) {
        double l2 = wlUm * wlUm;
        double n2 = 1
                + 0.6961663 * l2 / (l2 - Math.pow(0.0684043, 2))
                + 0.4079426 * l2 / (l2 - Math.pow(0.1162414, 2))
                + 0.8974794 * l2 / (l2 - Math.pow(9.896161, 2));
        return Math.sqrt(n2);
    }

    /** GeO2 glass Sellmeier index (Fleming 1984). */
    static double nGeO2(double wlUm) {
        double l2 = wlUm * wlUm;
        double n2 = 1
                + 0.80686642 * l2 / (l2 - Math.pow(0.068972606, 2))
                + 0.71815848 * l2 / (l2 - Math.pow(0.15396605, 2))
                + 0.85416831 * l2 / (l2 - Math.pow(11.841931, 2));
        return Math.sqrt(n2);
    }

    /** Lorentz-Lorentz molar refraction term. */
    static double lorentz(double n) {
        return (n * n - 1) / (n * n + 2);
    }

    /** Molar-averaged bulk index of (1-x) SiO2 + x GeO2. */
    static double nBulk(double xGeO2, double wlUm) {
        double Rm = (1 - xGeO2) * lorentz(nSiO2(wlUm)) + xGeO2 * lorentz(nGeO2(wlUm));
        return Math.sqrt((1 + 2 * Rm) / (1 - Rm));
    }

    /** Bulk density [g/cm3], linear in mol fraction. */
    static double rhoBulk(double xGeO2) {
        return (1 - xGeO2) * 2.20 + xGeO2 * 3.65;
    }

    // ---- Method A ----
    static double packingFromIndex(double nFilm, double xGeO2, double wlUm) {
        return lorentz(nFilm) / lorentz(nBulk(xGeO2, wlUm));
    }

    // ---- Method B ----
    static double packingFromDensity(double rhoFilm, double xGeO2) {
        return rhoFilm / rhoBulk(xGeO2);
    }

    // ---- Method C ----
    static double packingFromDeposition(double sih4, double n2o, double geh4,
                                        double tempC, double pTorr, double rfW) {
        double p0 = 0.94, Ea = 0.15, k = 8.617e-5, Tref = 573.15;
        double Tk = tempC + 273.15;
        double fT = Math.exp(-(Ea / k) * (1 / Tk - 1 / Tref));
        double rOx = n2o / sih4;
        double fOx = 1 - 0.08 * Math.exp(-rOx / 15);
        double fIon = 1 + 0.015 * Math.log(1 + (rfW / pTorr) / 150);
        return Math.min(p0 * fT * fOx * fIon, 1.02);
    }

    public static void main(String[] args) {
        double pA = packingFromIndex(1.448, 0.05, 1.55);
        System.out.printf("A. Lorentz-Lorentz : n=1.448 x=0.05  n_bulk=%.4f  p=%.4f%n",
                nBulk(0.05, 1.55), pA);

        double pB = packingFromDensity(2.15, 0.05);
        System.out.printf("B. Density ratio    : rho=2.15 x=0.05  rho_bulk=%.3f  p=%.4f%n",
                rhoBulk(0.05), pB);

        double pC = packingFromDeposition(20, 900, 5, 300, 1.0, 150);
        System.out.printf("C. Deposition        : T=300C  p=%.4f%n", pC);
    }
}

File: public/pecvd-packing-density/packing_density.py
#!/usr/bin/env python3
"""PECVD Ge-doped SiO2 packing density — reference implementation.

Three independent estimators (matching the web page):
  A. Lorentz-Lorentz    : from measured refractive index n
  B. Density ratio       : p = rho_film / rho_bulk
  C. Deposition-condition: empirical Arrhenius/oxidation/ion model

Packing density p = (film molar density) / (fully-dense bulk).
p < 1 indicates porosity / voids / incomplete network (e.g. Si-OH, Si-H).
"""

import math


# ---------- Sellmeier bulk indices ----------
def n_sio2(wl_um: float) -> float:
    """Fused silica, Malitson 1965."""
    l2 = wl_um * wl_um
    n2 = (1
          + 0.6961663 * l2 / (l2 - 0.0684043**2)
          + 0.4079426 * l2 / (l2 - 0.1162414**2)
          + 0.8974794 * l2 / (l2 - 9.896161**2))
    return math.sqrt(n2)


def n_geo2(wl_um: float) -> float:
    """GeO2 glass, Fleming 1984."""
    l2 = wl_um * wl_um
    n2 = (1
          + 0.80686642 * l2 / (l2 - 0.068972606**2)
          + 0.71815848 * l2 / (l2 - 0.15396605**2)
          + 0.85416831 * l2 / (l2 - 11.841931**2))
    return math.sqrt(n2)


def _R(n: float) -> float:
    """Lorentz-Lorentz molar refraction term."""
    return (n * n - 1) / (n * n + 2)


def n_bulk(x_geo2: float, wl_um: float) -> float:
    """Molar-averaged bulk index of (1-x) SiO2 + x GeO2."""
    Rm = (1 - x_geo2) * _R(n_sio2(wl_um)) + x_geo2 * _R(n_geo2(wl_um))
    return math.sqrt((1 + 2 * Rm) / (1 - Rm))


def rho_bulk(x_geo2: float) -> float:
    """Bulk density [g/cm3], linear in mol fraction (SiO2 2.20, GeO2 3.65)."""
    return (1 - x_geo2) * 2.20 + x_geo2 * 3.65


# ---------- Method A: Lorentz-Lorentz ----------
def packing_from_index(n_film: float, x_geo2: float = 0.0, wl_um: float = 1.55) -> float:
    nb = n_bulk(x_geo2, wl_um)
    return _R(n_film) / _R(nb)


# ---------- Method B: density ratio ----------
def packing_from_density(rho_film: float, x_geo2: float = 0.0) -> float:
    return rho_film / rho_bulk(x_geo2)


# ---------- Method C: deposition-condition empirical ----------
def packing_from_deposition(sih4_sccm: float, n2o_sccm: float, geh4_sccm: float,
                            temp_c: float, pressure_torr: float, rf_w: float):
    """Empirical densification model. CALIBRATE constants to your reactor.

    p = p0 * f_T * f_ox * f_ion
      f_T   = exp[-(Ea/k)(1/T - 1/Tref)]     temperature (Arrhenius)
      f_ox  = 1 - A*exp(-r_ox/r0)            oxidant ratio N2O/SiH4
      f_ion = 1 + B*ln(1 + (RF/P)/S0)        ion-bombardment energy
    """
    p0, Ea, k, Tref = 0.94, 0.15, 8.617e-5, 573.15  # 300 C reference
    Tk = temp_c + 273.15
    f_T = math.exp(-(Ea / k) * (1 / Tk - 1 / Tref))
    r_ox = n2o_sccm / sih4_sccm
    f_ox = 1 - 0.08 * math.exp(-r_ox / 15)
    f_ion = 1 + 0.015 * math.log(1 + (rf_w / pressure_torr) / 150)
    p = min(p0 * f_T * f_ox * f_ion, 1.02)
    x_ge = 0.6 * geh4_sccm / (sih4_sccm + geh4_sccm)   # incorporation eff ~0.6
    return {"p": p, "f_T": f_T, "f_ox": f_ox, "f_ion": f_ion,
            "r_ox": r_ox, "x_ge": x_ge}


if __name__ == "__main__":
    print("A. Lorentz-Lorentz : n=1.470, x=0.05, 1.55um")
    pA = packing_from_index(1.470, 0.05, 1.55)
    print(f"   n_bulk={n_bulk(0.05,1.55):.4f}  p={pA:.4f}  void={(1-pA)*100:.2f}%\n")

    print("B. Density ratio    : rho_film=2.15, x=0.05")
    pB = packing_from_density(2.15, 0.05)
    print(f"   rho_bulk={rho_bulk(0.05):.3f}  p={pB:.4f}  void={(1-pB)*100:.2f}%\n")

    print("C. Deposition        : SiH4=20 N2O=900 GeH4=5 T=300C P=1.0Torr RF=150W")
    r = packing_from_deposition(20, 900, 5, 300, 1.0, 150)
    print(f"   r_ox={r['r_ox']:.1f}  f_T={r['f_T']:.4f}  f_ox={r['f_ox']:.4f}  "
          f"f_ion={r['f_ion']:.4f}")
    print(f"   x_Ge={r['x_ge']:.4f}  p={r['p']:.4f}  void={(1-r['p'])*100:.2f}%")
